File: sort/shell_sort/Kotlin/ShellSort.kt
fun ShellSort(arr: IntArray): Int {
    val sz=arr.size
    var gap = sz/2
    
    while (gap>0) {
        var i = gap
        while (i<sz) {
            val temp = arr[i]
			var j = i
            while (j>=gap && (arr[j-gap]>temp)) {
                arr[j] = arr[j-gap]
                j-=gap
            }
            arr[j]=temp
            i += 1
        }
        gap /= 2
    }
    return 0
}

fun main(args: Array<String>) {
    val arr = intArrayOf(65, 83, 21, 78, 55, 12, 9, 33)
    println("Before Sorting")
    for( a in arr) print("$a  ")

    ShellSort(arr)

    println("\nAfter sorting")
    for( a in arr) print("$a  ")
}

File: data_structures/sorting/ShellSort.cpp
#include <iostream>
using namespace std;

int shellSort(int arr[], int n)
{
    for (int gap = n / 2; gap > 0; gap /= 2)
    {
        for (int i = gap; i < n; i += 1)
        {
            int temp = arr[i];

            int j;
            for (j = i; j >= gap && arr[j - gap] > temp; j -= gap)
                arr[j] = arr[j - gap];

            arr[j] = temp;
        }
    }
    return 0;
}

void printArray(int arr[], int n)
{
    for (int i = 0; i < n; i++)
        cout << arr[i] << " ";
}

int main()
{
    int arr[] = {12, 34, 54, 2, 3}, i;
    int n = sizeof(arr) / sizeof(arr[0]);

    cout << "Array before sorting: \n";
    printArray(arr, n);

    shellSort(arr, n);

    cout << "\nArray after sorting: \n";
    printArray(arr, n);

    return 0;
}
